localstack-core/localstack/services/lambda_/provider.py

```diff
@@ -1290,6 +1259,13 @@ def delete_function(
         function_name, qualifier = api_utils.get_name_and_qualifier(
             function_name, qualifier, context
         )
+
+        if qualifier and qualifier.startswith("-"):
```

Comment:
> Qualifiers do not need to start with a "-".

Precisely! If a qualifier/alias does start with a `-` then AWS interprets this as a user trying to delete multiple aliases and throws an error. See:

https://github.com/localstack/localstack/blob/2979749c47205e000aa3315e6a9ea9c0980f1ad2/localstack-core/localstack/services/lambda_/provider.py#L1263-L1267


> We should check if the qualifier is an alias, and then not allow it.

The `get_name_and_qualifier` checks this and assigns to the `qualifier` variable a qualifier derived from an ARN/function-name OR a qualifier passed in as a standalone parameter: 

See below how `qualifier` is re-assigned:
https://github.com/localstack/localstack/blob/2979749c47205e000aa3315e6a9ea9c0980f1ad2/localstack-core/localstack/services/lambda_/provider.py#L1259-L1261

See AWS and LocalStack output (with the prefix check):
```
$ aws lambda delete-function --function-name "hello-world" --qualifier -1
An error occurred (InvalidParameterValueException) when calling the DeleteFunction operation: Deletion of aliases is not currently supported.

$ aws lambda delete-function --function-name "hello-world:-1"               
An error occurred (InvalidParameterValueException) when calling the DeleteFunction operation: Deletion of aliases is not currently supported.

$ awslocal lambda delete-function --function-name "hello-world:-1"
An error occurred (InvalidParameterValueException) when calling the DeleteFunction operation: Deletion of aliases is not currently supported.

$ awslocal lambda delete-function --function-name "hello-world" --qualifier -1
An error occurred (InvalidParameterValueException) when calling the DeleteFunction operation: Deletion of aliases is not currently supported.
```

Reply:
I am aware of the qualfier checks and assesment, but this is not what I am talking about:

You cannot run `delete-function` with a qualifier pointing to an alias, only to a version:
```
$ aws lambda delete-function --function-name "hello-world" --qualifier test
An error occurred (InvalidParameterValueException) when calling the DeleteFunction operation: Deletion of aliases is not currently supported.
```

It does not matter if the alias starts with an `-`, and it does not interpret it as multiple aliases. you just cannot delete a function version by alias.

What is indeed valid, is something like this:

```
$ aws lambda delete-function --function-name "hello-world" --qualifier 1
```

This would delete only function version `1` of the `hello-world` function - not the entire function. However, the qualifier cannot be an alias, and an alias does not have to start with a dash. Instead of checking if it starts with a dash, you should use `api_utils.qualifier_is_alias` for the check if the qualifier is an alias.

Reply:
A-ha! I see what you're saying. That makes sense.

Reply:
Could we get this change in as well, before I review? :)

Reply:
Done! Thinking we can also probably get rid of some of these `TODOs` now above `delete_function`:


- TODO: does deleting the latest published version affect the next versions number?
- ~TODO: what happens when we call this with a qualifier and a fully qualified ARN? (+ conflicts?)~
- ~TODO: test different ARN patterns (shorthand ARN?)~
- TODO: test deleting across regions?
- TODO: test mismatch between context region and region in ARN
- ~TODO: test qualifier $LATEST, alias-name and version~